module/executiondatasync/requester/fulfiller.go

```diff
@@ -0,0 +1,199 @@
+package requester
+
+import (
+	"fmt"
+
+	"github.com/rs/zerolog"
+
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module"
+	"github.com/onflow/flow-go/module/component"
+	"github.com/onflow/flow-go/module/executiondatasync/execution_data"
+	"github.com/onflow/flow-go/module/executiondatasync/tracker"
+	"github.com/onflow/flow-go/module/irrecoverable"
+	"github.com/onflow/flow-go/module/util"
+)
+
+type jobResult struct {
+	executionData *execution_data.BlockExecutionData
+	resultID      flow.Identifier
+	blockHeight   uint64
+	err           error
+}
+
+type fulfiller struct {
+	jobResults    map[uint64]map[flow.Identifier]*jobResult
+	sealedResults map[uint64]flow.Identifier
+	fulfilled     map[uint64]*jobResult
+
+	fulfilledHeight uint64
+	sealedHeight    uint64
+
+	jobsIn     chan<- interface{}
+	jobsOut    <-chan interface{}
+	resultsIn  chan<- interface{}
+	resultsOut <-chan interface{}
+
+	logger  zerolog.Logger
+	metrics module.ExecutionDataRequesterMetrics
+
+	notifier *notifier
+	storage  tracker.Storage
+
+	component.Component
+}
+
+func newFulfiller(
+	fulfilledHeight uint64,
+	notifier *notifier,
+	storage tracker.Storage,
+	logger zerolog.Logger,
+	metrics module.ExecutionDataRequesterMetrics,
+) *fulfiller {
+	jobsIn, jobsOut := util.UnboundedChannel()
+	resultsIn, resultsOut := util.UnboundedChannel()
+
+	f := &fulfiller{
+		jobResults:      make(map[uint64]map[flow.Identifier]*jobResult),
+		sealedResults:   make(map[uint64]flow.Identifier),
+		fulfilled:       make(map[uint64]*jobResult),
+		jobsIn:          jobsIn,
+		jobsOut:         jobsOut,
+		resultsIn:       resultsIn,
+		resultsOut:      resultsOut,
+		fulfilledHeight: fulfilledHeight,
+		sealedHeight:    fulfilledHeight,
+		notifier:        notifier,
+		storage:         storage,
+		logger:          logger.With().Str("subcomponent", "fulfiller").Logger(),
+		metrics:         metrics,
+	}
+
+	cm := component.NewComponentManagerBuilder().
+		AddWorker(f.loop).
+		Build()
+
+	f.Component = cm
+
+	return f
+}
+
+func (f *fulfiller) submitJobResult(j *jobResult) {
+	f.jobsIn <- j
+}
+
+func (f *fulfiller) submitSealedResult(resultID flow.Identifier) {
+	f.resultsIn <- resultID
+}
+
+func (f *fulfiller) fulfill() error {
+loop:
+	if j, ok := f.fulfilled[f.fulfilledHeight+1]; ok {
+		f.fulfilledHeight++
+		delete(f.fulfilled, f.fulfilledHeight)
+		delete(f.sealedResults, f.fulfilledHeight)
+		f.notifier.notify(j.blockHeight, j.executionData)
+
+		goto loop
+	}
```

Comment:
```suggestion
	for {
		job, ok := f.fulfilled[f.fulfilledHeight+1]
		if !ok {
			break
		}
		
		f.fulfilledHeight++
		delete(f.fulfilled, f.fulfilledHeight)
		delete(f.sealedResults, f.fulfilledHeight)
		f.notifier.notify(job.blockHeight, job.executionData)
	}
```
Better to use the more common loop style.